lean4-infoview/src/infoview/info.tsx

```diff
@@ -281,137 +268,174 @@ export type InfoProps = InfoPinnable & { pos?: DocumentPosition };
+            .then(diags => diags.length === 0 ? lspDiagsHere.map(lspDiagToInteractive) : diags)
 
         // While `lake print-paths` is running, the output of Lake is shown as
         // info diagnostics on line 1.  However, all RPC requests block until
         // Lake is finished, so we don't see these diagnostics while Lake is
         // building.  Therefore we show the LSP diagnostics on line 1 if the
         // server does not respond within half a second.
         if (pos.line === 0 && lspDiagsHere.length) {
-            setTimeout(() => {
-                if (tickAtStart > displayPropsTick.current) {
-                    setDisplayProps({ pos, messages: lspDiagsHere, rpcSess, status: 'updating' })
-                    displayPropsTick.current = tickAtStart
-                }
-            }, 500)
+            setTimeout(() =>
+                resolve([undefined, undefined, undefined, lspDiagsHere.map(lspDiagToInteractive)]),
+                500)
         }
 
-        let newProps: Omit<InfoDisplayProps, 'triggerUpdate'>
-        try {
-            // NB: it is important to await both reqs at once, otherwise
-            // if both throw then one exception becomes unhandled.
-            const [goals, termGoal, userWidgets, messages] = await Promise.all([goalsReq, termGoalReq, widgetsReq, messagesReq]);
-            newProps = { pos, messages, goals, termGoal, userWidgets, rpcSess, status: 'ready' }
-        } catch (ex: any) {
-            if (ex?.code === RpcErrorCode.ContentModified) {
-                // Document has been changed since we made the request, try again
-                return void triggerUpdate();
-            } else if (ex?.code === RpcErrorCode.RpcNeedsReconnect) {
-                // Need to reconnect to RPC session
-                return void triggerUpdate();
+        // NB: it is important to await await reqs at once, otherwise
+        // if both throw then one exception becomes unhandled.
+        Promise.all([goalsReq, termGoalReq, widgetsReq, messagesReq]).then(
+            val => resolve(val),
+            ex => {
+                if (ex?.code === RpcErrorCode.ContentModified ||
+                    ex?.code === RpcErrorCode.RpcNeedsReconnect) {
+                    // Document has been changed since we made the request, or we need to reconnect
+                    // to the RPC sessions. Try again.
+                    setUpdaterTick(t => t + 1)
+                    reject('retry')
+                }
+                reject(ex)
             }
-
-            let errorString : string;
-            if (typeof ex === 'string') {
-                errorString = ex
-            } else if (isRpcError(ex)) {
-                errorString = mapRpcError(ex).message
-            } else if (ex instanceof Error) {
-                errorString = ex.toString()
+        )
+    }), [updaterTick, pos.uri, pos.line, pos.character, serverIsProcessing, lspDiagsHere])
```

Comment:
This throws React into an infinite loop. It is because `RpcSessionForFile.at(pos)` [creates a new object](https://github.com/leanprover/vscode-lean4/blob/1099571f9d44a3d5672e021d90d1886bf446c5e1/lean4-infoview-api/src/rpcSessions.ts#L159). This makes `useAsyncWithTrigger` decide 'oh, deps changed, let's set `asyncState.state` to `'notStarted'`. Then the effect fires in response and calls `triggerUpdate`. This makes the state change to `'loading'`. But oh look, on re-render we call `useRpcSessionAtPos` which produces a new object and we end up back at `asyncState.state === 'notStarted'`, repeat.

In general, the lack of a stable object identity in `useRpcSessionAtPos` makes it not really usable as a React dep. This looks like a bug so I will fix.